### DISCUSSION TITLE
How to setup msw on the client?

### DISCUSSION THREAD

**Discussion by jansedlon** [remix-run/remix#6624] [Q&A] [answered]
Hi, I followed few tutorials to setup server and client msw. The server side works great, but I cannot get to work the client side.

My `entry.client.tsx` looks like this

```tsx
async function prepareApp() {
	if (process.env.NODE_ENV === 'development') {
                // This file contains browser worker
		const { worker } = await import('~/utils/mocks.ts');
		return await worker.start();
	}

	return Promise.resolve();
}

prepareApp().then(() => {
	startTransition(() => {
		hydrateRoot(
			document,
			<StrictMode>
				<RemixBrowser />
			</StrictMode>,
		);
	});
});
```

The mock server starts, however it starts only after the document is loaded. So if I'm trying to mock a call to my CDN url with images, the image is loaded first and then the mock server starts. Which is unfortunate, because then it doesn't mock the CDN call to image.

Do you know how to fix it?

**Comment by sergiodxa** [comment, marked as answer]
You will have to delay loading the images until you have setup MSW, if you delay the hydration of React too after MSW is setup you could SSR your images with a placeholder and once hydrated replace them with the correct URL.

**Reply by jansedlon** [reply]
Hmmm, that's actually a good idea. So something like this? This function would be called on the client
```ts
function getImgSrc(src: string | null) {
 return process.env.NODE_ENV === 'development' && typeof window === 'undefined' ? null : src
}
```

**Reply by sergiodxa** [reply]
That will cause a hydration error, because the result of the function will be different on SSR and on hydration, checkout the [useHydrated hook from Remix Utils](https://github.com/sergiodxa/remix-utils#usehydrated), it handles that case by using a second render after hydration to switch to the hydrated state (where you will render the image), also [ClientOnly from Remix Utils](https://github.com/sergiodxa/remix-utils#clientonly) which gives you a JSX API to do the same.
